src/fix/main.rs

```diff
@@ -0,0 +1,954 @@
+    let mut bank = [0; BANK_SIZE]; // Temp buffer used to store a whole bank's worth of data
+    let mut global_sum = 0; // Global checksum variable
+
+    // Handle ROMX
+    /* input.as_raw_fd() == output.as_raw_fd() */
+    {
+        if file_size >= (0x10000 * BANK_SIZE) as u64 {
+            return Err(Error::TooManyBanks(file_size / BANK_SIZE as u64));
+        }
+        // Compute number of banks and ROMX len from file size
+        nb_banks = ((file_size + (BANK_SIZE - 1) as u64) / BANK_SIZE as u64) as u32;
+        total_romx_len = if file_size >= BANK_SIZE as u64 {
+            (file_size - BANK_SIZE as u64) as usize
+        } else {
+            0
+        };
+    } 
+
+    if let Some(pad_value) = options.pad_value {
+        // We want at least 2 banks
+        if nb_banks == 1 {
+            if rom0_len != rom0.len() {
+                // Fill the remaining space in rom0 with pad_value
+                for i in rom0.iter_mut().skip(rom0_len) {
+                    *i = pad_value;
+                }
+                // Update rom0Len to reflect the full size of rom0
+                rom0_len = rom0.len();
+            }
+            nb_banks = 2;
+        } else {
+            assert_eq!(rom0_len, rom0.len(), "rom0Len must equal the size of rom0");
+        }
+        assert!(nb_banks >= 2, "Number of banks must be at least 2");
+        // Alter number of banks to reflect required value
+        // x&(x-1) is zero iff x is a power of 2, or 0; we know for sure it's non-zero,
+        // so this is true (non-zero) when we don't have a power of 2
+        if !nb_banks.is_power_of_two() {
+            nb_banks = nb_banks.next_power_of_two();
+        }
+        // Write final ROM size
+        rom0[0x148] = (nb_banks / 2).trailing_zeros() as u8;
+        // Alter global checksum based on how many bytes will be added (not counting ROM0)
+        global_sum += pad_value.wrapping_mul(((nb_banks - 1) * BANK_SIZE as u32 - total_romx_len as u32) as u8) as u16;
+    }
+
+    // Handle the header checksum after the ROM size has been written
+    if options.fix_spec.header.is_some() {
+        let sum = rom0[0x134..0x14D]
+            .iter()
+            .fold(0u8, |a, i| a.wrapping_sub(i + 1));
+        overwrite_byte(
+            &mut rom0,
+            0x14D,
+            if matches!(options.fix_spec.header, Some(FixState::Trash)) {
+                !sum
+            } else {
+                sum
+            },
+            "header checksum",
+            options.overwrite,
+        );
+    }
+
+    if options.fix_spec.global.is_some() {
+        assert!(rom0_len >= 0x14E, "ROM0 length must be at least 0x14E");
+        for i in rom0[0..0x14E].iter().chain(rom0[0x150..rom0_len].iter()) {
+            global_sum = global_sum.wrapping_add(*i as u16);
+        }
+        // Pipes have already read ROMX and updated global_sum, but not regular files
+        /* if true || input.as_raw_fd() == output.as_raw_fd() */
```

Comment:
Why are those commented out?

Reply:
They are a leftover of when I wasn't sure on if the CLI would be able to receive two file arguments (an output and an input file, both separate). From what I understand, these checks are not as relevant in the Rust version.

These comments have been removed in my last commit.